scripts/leapp_preupgrade.py

```diff
@@ -0,0 +1,340 @@
+        os.remove(json_report_path)
+
+    text_report_path = "/var/log/leapp/leapp-report.txt"
+    if os.path.exists(text_report_path):
+        os.remove(text_report_path)
+
+
+def execute_preupgrade(command):
+    print("Executing preupgrade ...")
+    _, _ = run_subprocess(command)
+
+    # NOTE: we do not care about returncode because non-null always means actor error (or leapp error)
+    # if returncode:
+    #     print(
+    #         "The process leapp exited with code '%s' and output: %s\n"
+    #         % (returncode, output)
+    #     )
+    #     raise ProcessError(message="Leapp exited with code '%s'." % returncode)
+
+
+def parse_results(output):
+    print("Processing preupgrade results ...")
+
+    json_report_path = "/var/log/leapp/leapp-report.json"
+    report_json = "Not found"
+    message = "Can't open json report at " + json_report_path
+    alert = True
+
+    print("Reading JSON report")
+    if os.path.exists(json_report_path):
+        with open(json_report_path, mode="r") as handler:
+            report_json = json.load(handler)
+
+        # NOTE: with newer schema we will need to parse groups instead of flags
+        report_entries = report_json.get("entries", [])
+        inhibitor_count = len(
+            [entry for entry in report_entries if "inhibitor" in entry.get("flags")]
+        )
+        message = "Your system has %s inhibitors out of %s potential problems." % (
+            inhibitor_count,
+            len(report_entries),
+        )
+        alert = inhibitor_count > 0
+
+    output.report_json = report_json
+    output.alert = alert
+    output.message = message
+
+    print("Reading TXT report")
+    text_report_path = "/var/log/leapp/leapp-report.txt"
+    report_txt = "Not found"
+    if os.path.exists(text_report_path):
+        with open(json_report_path, mode="r") as handler:
+            report_txt = handler.read()
+
+    output.report = report_txt
+
+
+def call_insights_client():
+    print("Calling insight-client in background for immediate data collection.")
+    run_subprocess(["insights-client"], wait=False)
+    # NOTE: we do not care about returncode or output because we are not waiting for process to finish
+
+
+def main():
+    # Exit if not RHEL 7 or 8
+    dist, version = get_rhel_version()
+    if dist != "rhel" or exit_for_non_eligible_releases(version):
+        raise ProcessError(
+            message='Exiting because distribution="%s" and version="%s"'
+            % (dist, version)
+        )
+
+    output = OutputCollector()
+
+    try:
+        # Init variables
+        preupgrade_command = ["/usr/bin/leapp", "preupgrade", "--report-schema=1.1.0"]
+        use_no_rhsm = False
+        rhui_pkgs = []
+        if version.startswith("7"):
+            leapp_install_command = [
+                "yum",
+                "install",
+                "leapp-upgrade",
+                "-y",
+                "--enablerepo=rhel-7-server-extras-rpms",
```

Comment:
Not sure how to enable this repo on our AWS machines, since there is no RHSM installed there. Testing on beaker/testing-farm rn.

So on testing-farm we can install `leapp-upgrade` even without the repo, again does not know the about the repo without RHSM. When used without the `--enablerepo` part, looks like it's used successfully

On beaker `leapp-upgrade` or the repo is not found, RHSM problems again.

Just thinking how to test this on our end, and also how to make sure each customer is able to install this.

Reply:
My goal was to have the script and ansible same, because if this is problem then it is problem in ansible playbook as well - to me it makes sense to keep both yamls with same functionality, one is just implemented in python and one in ansible.

At least for now, till we sort out with insights how to even show the two different tasks in UI - currently there is no restriction on our tasks to show just 'some' systems eligible for that task.

Reply:
Yeah, but this is a bigger problem when it comes to rhel7... Will bring this up on our meetings.